src/ligero/verifier.rs

```diff
@@ -1,3 +1,273 @@
+use sha2::{Digest, Sha256};
+
+pub fn ligero_verify<FE: CodecFieldElement + LagrangePolynomialFieldElement>(
+    commitment: LigeroCommitment,
+    proof: &LigeroProof<FE>,
+    transcript: &mut Transcript,
+    linear_constraints: &LinearConstraints<FE>,
+    quadratic_constraints: &[QuadraticConstraint],
+    tableau_layout: &TableauLayout,
+) -> Result<(), anyhow::Error> {
+    // Write 0xdeadbeef, padded to 32 bytes, to the transcript to match what longfellow-zk does
+    transcript.write_byte_array(&[
+        0xde, 0xad, 0xbe, 0xef, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00,
+    ])?;
+
+    // The blind is also "u" in the specification. Generate one blind element for each witness
+    // and quadratic witness row in the tableau.
+    let low_degree_test_blind =
+        transcript.generate_challenge::<FE>(tableau_layout.num_constraint_rows())?;
+    let linear_constraint_alphas = transcript.generate_challenge::<FE>(linear_constraints.len())?;
+    let quad_constraint_alphas =
+        transcript.generate_challenge::<FE>(3 * quadratic_constraints.len())?;
+    let quad_proof_blinding =
+        transcript.generate_challenge::<FE>(tableau_layout.num_quadratic_triples())?;
+
+    transcript.write_field_element_array(&proof.low_degree_test_proof)?;
+    transcript.write_field_element_array(&proof.dot_proof)?;
+    transcript.write_field_element_array(&proof.quadratic_proof.0)?;
+    transcript.write_field_element_array(&proof.quadratic_proof.1)?;
+
+    let requested_column_indices = transcript.generate_naturals_without_replacement(
+        tableau_layout.num_columns() - tableau_layout.dblock(),
+        tableau_layout.num_requested_columns(),
+    );
+
+    // Check that dot product matches linear constraints
+    let want_dot_product = linear_constraints
+        .right_hand_side_terms()
+        .iter()
+        .zip(&linear_constraint_alphas)
+        .fold(FE::ZERO, |sum, (rhs_term, alpha)| sum + *rhs_term * alpha);
+    let proof_dot_product = proof
+        .dot_proof
+        .iter()
+        // Skip the nreq random values at the start of the row. The proof only sums over the
+        // witnesses.
+        // Not documented in the specification.
+        .skip(tableau_layout.num_requested_columns())
+        .take(tableau_layout.witnesses_per_row())
+        .fold(FE::ZERO, |sum, term| sum + term);
+    if want_dot_product != proof_dot_product {
+        return Err(anyhow!("dot product mismatch"));
+    }
+
+    let inner_product_vector = inner_product_vector(
+        tableau_layout,
+        linear_constraints,
+        &linear_constraint_alphas,
+        quadratic_constraints,
+        &quad_constraint_alphas,
+    )?;
+
+    // Check that dot proof matches requested columns
+    let mut want_dot_row = proof.tableau_columns[TableauLayout::dot_proof_row()].clone();
```

Comment:
Here and beyond (87, 114, 139 and more) are array accesses that aren't checked beforehand. Like, `LigeroProof::decode()` validates that total element count but since it doesn't first verify the `Vec<Vec<...>>` structure, seems like this is a DoS from a malicious input.

Consider doing some preflight checks at the top of this, something like:

```
  if proof.tableau_columns.len() != tableau_layout.num_rows()
      return Err

  for row in proof.tableau_columns.iter()
      if row.len() != tableau_layout.num_requested_columns()
          return Err

  if proof.merkle_tree_nonces.len() != tableau_layout.num_requested_columns 
       return Err
```

Reply:
Preflighting the validity of the proof is a good idea, but I think I'd rather do it in `LigeroProof::decode` and only do it once. My rationale is that you want to do that kind of safety check on untrusted data, which would be messages you're serializing off the wire. If you constructed a `LigeroProof` in code and it's invalid, then it seems OK for this verify function to panic because it accesses past the end of the array because, well, don't do that. What do you think?